File: Python/intersection-of-three-sorted-arrays.py
# Time:  O(n)
# Space: O(1)

class Solution(object):
    def arraysIntersection(self, arr1, arr2, arr3):
        """
        :type arr1: List[int]
        :type arr2: List[int]
        :type arr3: List[int]
        :rtype: List[int]
        """
        result = []
        i, j, k = 0, 0, 0
        while i != len(arr1) and j != len(arr2) and k != len(arr3):
            if arr1[i] == arr2[j] == arr3[k]:
                result.append(arr1[i])
                i += 1
                j += 1
                k += 1
            else:
                curr = max(arr1[i], arr2[j], arr3[k])
                while i != len(arr1) and arr1[i] < curr:
                    i += 1
                while j != len(arr2) and arr2[j] < curr:
                    j += 1
                while k != len(arr3) and arr3[k] < curr:
                    k += 1
        return result


# Time:  O(n)
# Space: O(n)
class Solution2(object):
    def arraysIntersection(self, arr1, arr2, arr3):
        """
        :type arr1: List[int]
        :type arr2: List[int]
        :type arr3: List[int]
        :rtype: List[int]
        """
        intersect = reduce(set.intersection, map(set, [arr2, arr3]))
        return [x for x in arr1 if x in intersect]

File: C++/intersection-of-three-sorted-arrays.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    vector<int> arraysIntersection(vector<int>& arr1, vector<int>& arr2, vector<int>& arr3) {
        vector<int> result;
        int i = 0, j = 0, k = 0;
        while (i != arr1.size() && j != arr2.size() && k != arr3.size()) {
            if (arr1[i] == arr2[j] && arr2[j] == arr3[k]) {
                result.emplace_back(arr1[i]);
                ++i, ++j, ++k;
            } else {
                int curr = max(max(arr1[i], arr2[j]), arr3[k]);
                for (; i != arr1.size() && arr1[i] < curr; ++i);
                for (; j != arr2.size() && arr2[j] < curr; ++j);
                for (; k != arr3.size() && arr3[k] < curr; ++k);
            }
        }
        return result;
    }
};
